### DISCUSSION TITLE
Expose code-server without root permission in remote server !

### DISCUSSION THREAD

**Discussion by duckhaitong** [coder/code-server#4023] [Q&A] [answered]
Hi everyone, I'm exposing code-server to my local machine. I've already set it up in my remote server but I can not use port forwarding or sudo command. How can I do that without root permission, thanks in advance.

**Comment by jsjoeio** [comment, marked as answer]
Hmm... Could you use something like ngrok to create a tunnel?

https://ngrok.com/

**Reply by duckhaitong** [reply]
I could not, even port fowarding.

**Reply by duckhaitong** [reply]
Can I you self-signed certificate approach without sudo command ?

**Reply by jsjoeio** [reply]
Oh I'm very sorry, I overlooked your comment "I can not use port forwarding." 

> Can I you self-signed certificate approach without sudo command ?

Based on what I'm seeing [in the docs](https://coder.com/docs/code-server/v3.11.1/guide#using-a-self-signed-certificate), I don't think so 🤔 But maybe @code-asher knows

**Reply by duckhaitong** [reply]
I just realized that I can use it but only with ports which are > 1024. eg 0.0.0.0:8080. Now I'm tryna set up https. I will update if I encounter new issues. by the way, thank you @jsjoeio

**Reply by jsjoeio** [reply]
Hooray! Sounds good

**Reply by AyaanZaveri** [reply]
You can also try [localtunnel](https://github.com/localtunnel/localtunnel) for tunnelling, you can create unlimited HTTP tunnels with custom subdomains.

**Installation**
npm i localtunnel -g

https://www.npmjs.com/package/localtunnel

**Reply by jsjoeio** [reply]
@AyaanZaveri I've never used localtunnel 👀 what's the difference between that and `ngrok`?

**Reply by AyaanZaveri** [reply]
I find localtunnel easier to use and install, although ngrok has more features like tcp. If you want to expose let's say an HTML file, all you would have to type is lt --port YOUR_PORT. Also, if you want to expose ports you can do it with localtunnel since there is no limit. There isn't too much of a difference though.

**Reply by duckhaitong** [reply]
Thanks @AyaanZaveri. I'm just able to use "pure" code-server, could not install anything else. I've changed cert in config.yaml to true, but I found no randomly generated self-signed certificate. Does It locate in ~/.local/share/code-server ?

**Reply by AyaanZaveri** [reply]
To install localtunnel you need to open terminal and type, `npm i localtunnel -g`. You don't have to change any code-server settings.

**Reply by jsjoeio** [reply]
> I find localtunnel easier to use and install, although ngrok has more features like tcp. If you want to expose let's say an HTML file, all you would have to type is lt --port YOUR_PORT. Also, if you want to expose ports you can do it with localtunnel since there is no limit. There isn't too much of a difference though.

Ahh, I see. That sounds super easy. I'll have to give a try next time I reach for ngrok! Thanks for the explanation!

**Reply by AyaanZaveri** [reply]
No problem.
